python/lsst/analysis/drp/gatherResourceUsage.py

```diff
@@ -0,0 +1,550 @@
+
+    Its output table has columns for each of the dimensions of the input
+    metadata's data ID, as well as (subject to configuration):
+
+    - ``memory``: the maximum resident set size for the entire quantum
+      (in bytes);
+    - ``prep_time``: the time spent in the pre-initialization step in
+      which the middleware checks which of the quantum's inputs are available;
+    - ``init_time``: the time spent in task construction;
+    - ``run_time``: the time spent executing the task's runQuantum
+      method.
+    - ``{method}``: the time spent in a particular task or subtask
+      method decorated with `lsst.utils.timer.timeMethod`.
+
+    All time durations are CPU times in seconds, and all columns are 64-bit
+    floating point.  Methods or steps that did not run are given a duration of
+    zero.
+
+    It is expected that this task will be configured to run multiple times in
+    most pipelines, often once for each other task in the pipeline.
+    """
+
+    ConfigClass = GatherResourceUsageConfig
+    _DefaultName = "gatherResourceUsage"
+
+    @classmethod
+    def build_quantum_graph(cls, metadata_refs):
+        """Build a specialized `QuantumGraph` that configures and runs this
+        task on existing metadata datasets.
+
+        Parameters
+        ----------
+        metadata_refs : `Iterable` [ `DatasetRef` ]
+            References to metadata datasets.  Non-metadata datasets are
+            silently ignored, in order to make it easier to pass in more
+            general dataset query results.
+
+        Notes
+        -----
+        This task cannot easily be added to a regular pipeline, as it's much
+        more natural to have it run automatically on all *other* tasks.
+        But even machine-generating a pipeline is problematic, because our
+        current `QuantumGraph` generation algorithm isn't smart enough to
+        recognize that those pipelines are usually disconnected subgraphs, and
+        that generating a graph for each of those separately is many orders of
+        magnitude faster (and *not* doing is disastrously slow).  As a
+        short-term workaround, this method can be used to generate a
+        `QuantumGraph` for just this task much more efficiently from the
+        metadata datasets already present in a set of collections.
+        """
+        # Group input metadata datasets by dataset type.
+        metadata_refs_by_dataset_type = defaultdict(set)
+        for ref in metadata_refs:
+            metadata_refs_by_dataset_type[ref.datasetType].add(ref)
+        # Iterate over those groups, creating a configuration of
+        # this task and quanta for each one.
+        quanta_by_task_def = {}
+        for input_metadata_dataset_type, metadata_refs in metadata_refs_by_dataset_type.items():
+            if not input_metadata_dataset_type.name.endswith("_metadata"):
+                continue
+            input_task_label = input_metadata_dataset_type.name[:-len("_metadata")]
```

Comment:
consider doing something like:
```suggestion
            input_task_label = input_metadata_dataset_type.name.replace("_metadata",'.')
```
or if you are afraid there is a `_metadata` somewhere else in the name, possibly regular expression or something. Its not bad persay like this, but it always feels off and weird looking in code to see strings used as sequences (although they are)

Reply:
I'll do the regex; I'm not really bothered by the indexing (I consider that different from `Sequence`-ness, though I'm not sure I can defend that distinction in my brain), but rolling it into the check in the previous couple of lines seems to make sense.